### PR TITLE
[http] Initial contribution

### PR DESCRIPTION
**Pull request by J-N-K**
This adds a new HTTP binding as a replacement for the old http1 binding.

Replaces #7851

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
PR build asks you to apply spotless to the pom.xml.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/GenericItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/GenericItemConverter.java:49]: I don't know if generics would complicate things. But if the compiler will accept `StringType::new` as a `Function<String, State>` instance then it should be possible to replace the following without difficulty.

```suggestion
public class GenericItemConverter extends AbstractTransformingItemConverter {
    private final Function<String, State> toStateFunction;

    public GenericItemConverter(Function<String, State> toStateFunction, Consumer<State> updateState, Consumer<Command> postCommand,
            @Nullable Consumer<String> sendHttpValue, ValueTransformation stateTransformations,
            ValueTransformation commandTransformations, HttpChannelConfig channelConfig) {
        super(updateState, postCommand, sendHttpValue, stateTransformations, commandTransformations, channelConfig);
        this.toStateFunction = toStateFunction;
    }

    protected State toState(String value) {
        return toStateFunction.apply(value);
    }
```

File: bundles/pom.xml
Comment (cpmeister) [inline bundles/pom.xml:122]: ?

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java:71]: So `result.isFailed()` returns false on a 401?

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.http/src/main/feature/feature.xml
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/feature/feature.xml:3]: According to the discussions in another PR: no.

Comment (Hilbrand) [reply bundles/org.openhab.binding.http/src/main/feature/feature.xml:3]: https://github.com/openhab/openhab-core/pull/1673

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java:76]: I went ahead and rearranged the logic here into something a bit more straight forward. I noticed that you don't handle other response types here which means that `future` might hang indefinitely if say a `504` occurs. You need a default fallback that completes the future regardless.
```suggestion
        if(result.isFailed()){
            logger.warn("Requesting '{}' (method='{}', content='{}') failed: {}", request.getURI(),
                        request.getMethod(), request.getContent(), result.getFailure().getMessage());
            future.complete(null);
        }else switch(response.getStatus()) {
            case HttpStatus.OK_200:
                byte[] content = getContent();
                String encoding = getEncoding();
                if (content != null) {
                    future.complete(new Content(content, encoding == null ? fallbackEncoding : encoding, getMediaType()));
                } else {
                    future.complete(null);
                }
                break;
            case HttpStatus.UNAUTHORIZED_401:
                logger.debug("Requesting '{}' (method='{}', content='{}') failed: Authorization error",
                        request.getURI(), request.getMethod(), request.getContent());
                future.completeExceptionally(new IllegalStateException("Auth Error"));
                break;
        }
```

**Review by cpmeister — CHANGES_REQUESTED** [review]

**Review by cpmeister — APPROVED** [review]

**Comment by J-N-K** [comment]
@openhab/add-ons-maintainers Can we get a second review here? Would be great to have this in M3, so people using http1 can move on to OH3

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java
Comment (fwolter) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java:51]: Does this annotation has any effect?

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java
Comment (fwolter) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java:80]: `info` is not adequate here. See https://www.openhab.org/docs/developer/guidelines.html#f-logging

"info logging should be used sparsely. e.g. a newly started component or a user file that has been loaded."

File: bundles/org.openhab.binding.http/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.http/src/main/resources/OH-INF/thing/thing-types.xml:8]:
```suggestion
		extensible="color,contact,datetime,dimmer,image,location,number,rollershutter,string,switch">
```

File: bundles/org.openhab.binding.http/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.http/src/main/resources/OH-INF/thing/thing-types.xml:16]: Should this be `url`?

File: bundles/org.openhab.binding.http/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.http/src/main/resources/OH-INF/thing/thing-types.xml:18]: `0` would raise an `IllegalArgumentException` in `scheduleWithFixedDelay()`.
```suggestion
			<parameter name="refresh" type="integer" unit="s" min="1">
```

**Review by fwolter — CHANGES_REQUESTED** [review]
Good code in general. I'm not aware of the current null check implementations, but I receive 3 warnings when compiling with maven.

I don't really like that you use unchecked exceptions all over the code, because the compiler doesn't enforce you to catch those. Normally I would claim this as error prone, but I think you know what you're doing.

File: bundles/org.openhab.binding.http/README.md
Comment (J-N-K) [reply bundles/org.openhab.binding.http/README.md:50]: If the response (after transformation) is equal to that value, it is interpreted as `ON` (e.g. a response could be `true`, `1` or `eingeschalltet`).

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java:176]: This is not written very often but an important information. So I think INFO is ok here. I adjusted the other message.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/Content.java
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/Content.java:42]: Null-Check is leftover from a previous version where encoding could be null.

IllegalCharsetNameException is a subset from IllegalArgumentException, so this also catechs an invalid name.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java:80]: Since this is an error state, I raised it to warn.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java:95]: `BigDecimal(string)` throws a NumberFormatException if `string` is not a number.

File: bundles/org.openhab.binding.http/README.md
Comment (J-N-K) [reply bundles/org.openhab.binding.http/README.md:120]: removed the section.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java:169]: But this only happens when the authentication is constructed. I clarified the message.

**Comment by J-N-K** [comment]
Regarding the checked/unchecked exceptions: Unfortunately an unchecked exception leads to a better code style.

`refreshingUrlCache.get().ifPresent(itemValueConverter::process);`

is not possible if `ItemValueConverter.process(Content content)` throws a check exception. The exception handling would need to be done insinde of `ifPresent`.

File: bundles/org.openhab.binding.http/README.md
Comment (fwolter) [reply bundles/org.openhab.binding.http/README.md:50]: Can you put this sentence in the readme? I think it would help the user to understand how this binding is actually working, as you don't provide examples.

When thinking about how this works, I also wouldn't know what to set the transformation parameters to. I think a full example would be really helpful.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.http/README.md
Comment (J-N-K) [reply bundles/org.openhab.binding.http/README.md:67]: I added a note for the special values and also examples for the transformation

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**
